python/dgl/runtime/scheduler.py

```diff
@@ -837,4 +916,75 @@ def _mfunc_wrapper(src_data, edge_data, dst_data):
     msg = ir.EDGE_UDF(_mfunc_wrapper, fdsrc, fdedge, fddst)
     return msg
 
+def _gen_send(graph, u, v, eid, mfunc, var_src_nf, var_dst_nf, var_ef):
+    """Internal function to generate send schedule"""
+    mfunc = _standardize_func_usage(mfunc, 'message')
+    mfunc_is_list = utils.is_iterable(mfunc)
+    # vars
+    var_u = var.IDX(u)
+    var_v = var.IDX(v)
+    var_eid = var.IDX(eid)
+
+    if mfunc_is_list:
+        if eid.is_slice(0, graph.number_of_edges()):
+            # full graph case
+            res = spmv.build_gidx_and_mapping_graph(graph)
+        else:
+            num_nodes = graph.number_of_nodes()
+            res = spmv.build_gidx_and_mapping_uv((u, v, eid), num_nodes)
+        adj, edge_map, _ = res
+        # create a tmp message frame
+        tmp_mfr = FrameRef(frame_like(graph._edge_frame._frame, len(eid)))
+        var_out = var.FEAT_DICT(data=tmp_mfr)
+        spmv.gen_v2e_spmv_schedule(graph=adj,
+                                   mfunc=mfunc,
+                                   src_frame=var_src_nf,
+                                   dst_frame=var_dst_nf,
+                                   edge_frame=var_ef,
+                                   out=var_out,
+                                   out_size=len(eid),
+                                   edge_map=edge_map)
+    else:
+        # UDF send
+        var_out = _gen_udf_send(graph, var_src_nf, var_dst_nf, var_ef, var_u,
+                                var_v, var_eid, mfunc)
+    return var_out
+
+def _build_idx_map(idx, nbits):
+    """Build a map from the input ids to continuous ids that starts from zero.
+    And the number of bits data type of each integer in the mapping uses will
+    be nbits
+
+    Examples
+    --------
+    >>> x = [1, 5, 3, 6]
+    >>> o2n = map_to_continuous(x)
+    >>> o2n
+    [n/a, 0, n/a, 2, n/a, 1, 3]
+
+    "n/a" will be filled with 0
+
+    Parameters
+    ----------
+    x : Index
+        The input ids, assumed to be unique.
+    nbits: int
+        Number of bits each integer in the mapping should use, can be 32 or 64
+
+    Returns
+    -------
+    old_to_new : CtxCachedObject
+        The mapping from old id to new id. It is a vector of length MAX(x).
+        One can use advanced indexing to convert an old id tensor to a
+        new id tensor: new_id = old_to_new[old_id]
+    """
+    x = idx.tousertensor()
+    map_len = int(F.asnumpy(F.max(x, dim=0))) + 1
+    old_to_new = F.zeros((map_len,), dtype=F.int64, ctx=F.cpu())
```

Comment:
Can we fill the n/a's with -1 since 0 is a valid value?  It's hard to debug otherwise.